test/util/list_files.js

```diff
@@ -0,0 +1,102 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert = require("chai").assert;
+const co     = require("co");
+
+const ListFiles       = require("../../lib/util/list_files");
+const RepoASTTestUtil = require("../../lib/util/repo_ast_test_util");
+
+describe("ListFiles", function () {
+    // Will always read "x".
+
+    const cases = {
+        "empty": {
+            state: "x=S:I README.md",
+            expected: [],
+        },
+        "file in meta": {
+            state: "x=S",
+            expected: ["README.md"],
+        },
+        "excluded by relative path in meta": {
+            state: "x=S",
+            relativePath: "foo",
+            expected: [],
+        },
+        "added file in meta": {
+            state: "x=S:I foo=bar",
+            expected: ["README.md", "foo"],
+        },
+        "deep file in meta": {
+            state: "x=S:I foo/bar/baz=bar",
+            expected: ["README.md", "foo/bar/baz"],
+        },
+        "filtered file in meta": {
+            state: "x=S:I foo/bar/baz=bar",
+            relativePath: "foo",
+            expected: ["bar/baz"],
+        },
+        "filtered deep file in meta": {
+            state: "x=S:I foo/bar/baz=bar",
+            relativePath: "foo/bar",
+            expected: ["baz"],
+        },
+        "with a closed sub": {
+            state: "a=S|x=U",
+            expected: ["README.md"],
+        },
+        "open sub": {
+            state: "a=S|x=U:Os",
+            expected: ["README.md", "s/README.md"],
+        },
+        "relative to sub": {
```

Comment:
We don't test the "relative to another sub" case.  That is, if you look relative to bar, do you *not* see foo/ files?  Also, these tests should probably check a per-sub filename, so that we know we're actually looking in the sub.